modules/libnetconf2-netconf-server@2023-09-07.yang

```diff
@@ -265,35 +265,33 @@ module libnetconf2-netconf-server {
       "Grouping for the SSH Keyboard interactive authentication method.";
 
     container keyboard-interactive {
-      presence "Indicates that PAM configuration file name has been configured.
-               This statement is present so the mandatory descendant
-               nodes do not imply that this node must be
-               configured.";
+      presence "Indicates that PAM configuration file name has been configured and that
```

Comment:
This description should be updated. Also, I propose to rename this container to something like `use-system-auth`.

Reply:
The proposed solution to this is to add a mandatory choice node to the presence container.

```
container keyboard-interactive {
    presence "";
    
    choice method {
        mandatory true;
        
        leaf pam-based {
            type empty;
        }
    }
}
```

This choice will, for now, only have a single case, that is `pam-based`. In the future as the library grows other cases may be added to the choice node allowing for different approaches when it comes to the keyboard interactive authentication.

The description of the `pam-based` leaf will reflect how it's handled in the library.

Reply:
I think this is suboptimal because it leaks some implementation details ("it's using PAM") into a configuration which, in my opinion, should just say "this is authenticating in the same way as the rest of the Linux system".

Another problem which we talked about with Michal over Slack the other day is that this concept allows different authentication backends to be used for the `password` and `keyboard-interactive` SSH authentication schemes, which makes no sense to me. We were not able to reach a mutual understanding about this point, though.

In short, I prefer the configuration to describe how the system behaves to an external observer. To a "NETCONF admin", the choice is really between "let me configure this explicitly defined list of users for this endpoint", which the IETF models are all about, vs. "delegate auth to the system for this endpoint", which I think is the purpose of this extension. In contrast to that, the choice between PAM and `getpwent`/`getspnam`/... is something that the system architect should be making. I cannot imagine that someone might want to provision a system where one half of SSH clients authenticate through the system's `/etc/shadow` while the rest goes through the system's PAM, especially since PAM's default config delegates everything to `/etc/shadow`, *and* when the decision is essentially only determined by the level of support of SSH features in the client' SSH library.

For me, the perfect approach will be a single, empty presence container `np2:user-system-auth` which, if present, will use PAM (with a PAM service name that's passed from a build-time-configurable default in Netopeer2 to libnetconf2 through an API call) for both kbd-interactive and password authentication. Or event better -- make it a non-presence container with a single leaf `enabled` so that I can add an augment which sets its default value to `true`. To me, this configuration is all about *policies*, not about fine-tuning library behavior in a level of detail that is, IMHO, not needed at all. The system is either *built with PAM*, which is a huge majority of any "classic" Linux systemd, or it is built *without PAM* which might be the case on some really restricted embedded boxes. Sure, have an `#ifdef` in code for that if you want to care about such systems, but using and bypassing PAM at the same time is something that I see no use case for.